rfcs/0003-timestamps-and-ttl.md

```diff
@@ -0,0 +1,295 @@
+    /// in separate layers of the LSM tree, you will see the following behavior:
+    /// - if retrieval_grace_ms is <=0, KeyA@100 will be returned, and only the first SST
+    ///   will be scanned
+    /// - if retrieval_grace_ms is <=50, KeyA@100 will be returned after having scanned
+    ///   the SST that contains KeyA@50
+    /// - if retrieval_grace_ms is > 50 and <= 100, KeyA@150 will be returned after having
+    ///   scanned the SST containing KeyA@150
+    /// - otherwise, KeyA@150 will be returned after scanning all levels of the LSM tree
+    ///
+    /// default: -1 (the first matching key will be retrieved)
+    pub retrieval_grace_ms: i64,
+}
+```
+
+**Callout for TTL**: I have decided that for this RFC we will not introduce the ability to disable
+filtering out TTL-expired records on retrieval. The default implementation will read the metadata
+for every retrieved row and filter out records where `ts + tll < now`. We can, as a future
+optimization, enable lazy deserialization of the metadata so that this doesn't need to happen on
+each row retrieved. This can be part of the API design for compaction filters (
+see https://github.com/slatedb/slatedb/issues/225)
+
+**Alternative**: if this retrieval grace concept is too complicated, we can also consider specifying
+a "base timestamp", that will retrieve the first result >= the configured base timestamp, or the 
+most recent one if none are greater than or equal to the base timestamp. This is significantly less
+flexible, however, as many rows may not be updated in a long time (though most of those rows will be
+in the base layer of the LSM tree anyway, so it may not be so bad).
+
+### SlateDB::Get
+
+SlateDB currently returns a `Option<Bytes>` on a successful read operation. Now that additional
+metadata (timestamps in this RFC) are introduced, we will need to expand this API to include
+more information in the return value. We have a few options here (TODO: update this section when
+we converge on the approach we want):
+
+1. We can return a strongly typed result (such as `Option<Row>` that is a struct with the result as
+   well as any additional retrieved metadata)
+2. We can introduce a `db#get_with_metadata` that will return a tuple of `Bytes` and corresponding
+   metadata
+3. Do nothing (make the timestamp opaque to the user and require them to interact with timestamps
+   only via the public API)
+
+## Implementation
+
+### Data Format
+
+Currently, SlateDB does not have a mechanism for backwards-compatibility of the SST data format.
+We will first introduce the following two fields to the flatbuffer definitions: 
+- `format_version`: field that specifies the codec version that encoded this SST
+- `meta_features`: an enumeration of metadata fields that are available with each encoded row. This
+  makes it possible in the future to save space per row by disabling metadata features, but that is
+  out of scope for this RFC
+
+```fbs
+enum MetaFeatures: byte {
```

Comment:
WYDT about including deletion tombstones as a metadata flag here? It would allow us to use the value len u32 as a normal u32 instead of reserving INT_MAX to signal a deletion (which is what we do now). I think it'd be less efficient since we'd need a byte per-row (I don't think FB supports a single bit in the schema).

Reply:
This makes sense to me. We're basically saying that if we're already including a `meta` encoded in each row, we may as well use 1 byte to encode that it's a tombstone instead and not encode the value length (saving 3 bytes total)

I think we could take this one step further and introduce a `ROW_TYPE` metadata flag, and `ROW_TYPE=0` would be an insertion (normal row) and `ROW_TYPE=1` would be a deletion. That way we could introduce future row types if they ever come up (not sure what the use case would be, but since fbs doesn't support single bits anyway we may as well leverage the entire byte in the format)